File: AndroidDev/+ Apps/TriviaQuiz/QuizApp/QuizGameActivity.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Xml;
using Android.App;
using Android.Content;
using Android.Graphics;
using Android.Graphics.Drawables;
using Android.OS;
using Android.Util;
using Android.Views;
using Android.Views.Animations;
using Android.Widget;
using DroidRes = Android.Resource;
using static QuizApp.QuizPreferences;

namespace QuizApp
{
   [Activity(Label = "@string/quiz_game_title")]
   public class QuizGameActivity : Activity
   {
      private readonly IDictionary<int, Question> _questions = new Dictionary<int, Question>(QuestionBatchSize);
      private ISharedPreferences _gameSettings;
      private Button _noButton;
      private ImageSwitcher _questionImageSwitcher;
      private TextSwitcher _questionTextSwitcher;
      private Button _yesButton;

      protected override void OnCreate(Bundle savedInstanceState)
      {
         base.OnCreate(savedInstanceState);
         SetContentView(Resource.Layout.Game);

         // Обработка кнопки Yes
         _yesButton = FindViewById<Button>(Resource.Id.Button_Yes);
         _yesButton.Click += (sender, args) => HandleAnswerAndShowNextQuestion(true);

         // Обработка кнопки No
         _noButton = FindViewById<Button>(Resource.Id.Button_No);
         _noButton.Click += (sender, args) => HandleAnswerAndShowNextQuestion();

         // Извлечение общих настроек приложения
         _gameSettings = GetSharedPreferences(GamePreferences, FileCreationMode.Private);

         // Загрузка вопросов
         var startQuestionNumber = _gameSettings.GetInt(GamePreferencesCurrentQuestion, default(int));

         // Если мы в только начинаем викторину, инициализируем общие настройкм
         if (startQuestionNumber == 0)
         {
            var editor = _gameSettings.Edit();
            editor.PutInt(GamePreferencesCurrentQuestion, 1).Commit();
            startQuestionNumber = 1;
         }

         LoadQuestionBatch(startQuestionNumber);

         var droidFadeIn = AnimationUtils.LoadAnimation(this, DroidRes.Animation.FadeIn);
         var droidFadeOut = AnimationUtils.LoadAnimation(this, DroidRes.Animation.FadeOut);

         // Установка Text Switcher
         _questionTextSwitcher = FindViewById<TextSwitcher>(Resource.Id.TextSwitcher_QuestionText);
         _questionTextSwitcher.InAnimation = droidFadeIn;
         _questionTextSwitcher.OutAnimation = droidFadeOut;
         // TODO: FIX THE _questionTextSwitcher.SetFactory(new QuizGameTextSwitcherFactory(this));         

         // Установка Image Switcher
         _questionImageSwitcher = FindViewById<ImageSwitcher>(Resource.Id.ImageSwitcher_QuestionImage);
         _questionImageSwitcher.InAnimation = droidFadeIn;
         _questionImageSwitcher.OutAnimation = droidFadeOut;
         // TODO: FIX THE _questionImageSwitcher.SetFactory(new QuizGameImageSwitcherFactory(this));

         if (_questions.ContainsKey(startQuestionNumber))
         {
            var questionText = GetQuestionText(startQuestionNumber);            
            _questionTextSwitcher.SetCurrentText(questionText);
            var questionImage = GetQuestionImage(startQuestionNumber);
            _questionImageSwitcher.SetImageDrawable(questionImage);
         }
         else
         {
            HandleNoQuestions();
         }
      }

      public override bool OnCreateOptionsMenu(IMenu menu)
      {
         base.OnCreateOptionsMenu(menu);

         MenuInflater.Inflate(Resource.Menu.GameOptions, menu);
         menu.FindItem(Resource.Id.help_menu_item).SetIntent(new Intent(this, typeof(QuizHelpActivity)));
         menu.FindItem(Resource.Id.settings_menu_item).SetIntent(new Intent(this, typeof(QuizSettingsActivity)));

         return true;
      }

      public override bool OnOptionsItemSelected(IMenuItem item)
      {
         base.OnOptionsItemSelected(item);
         StartActivity(item.Intent);
         return true;
      }

      /// <summary>
      ///    Вспомогательный метод, отвечающий за ситуацию отсутствия вопросов
      /// </summary>
      private void HandleNoQuestions()
      {
         _questionTextSwitcher?.SetText(Resources.GetText(Resource.String.no_questions));
         _questionImageSwitcher?.SetImageResource(Resource.Drawable.NoQuestion);
         _yesButton.Enabled = false;
         _noButton.Enabled = false;
      }

      /// <summary>
      ///    Извлекает объект Drawable для вопроса
      /// </summary>
      /// <param name="questionNumber">Номер вопроса</param>
      /// <returns>Объект Drawable</returns>
      private Drawable GetQuestionImage(int questionNumber)
      {
         Drawable image;

         try
         {
            using (var webClient = new WebClient())
            {
               var imageUri = new Uri(GetQuestionImageUri(questionNumber));
               var imageData = webClient.DownloadData(imageUri);
               using (var stream = new MemoryStream(imageData))
               {
                  var bitmap = BitmapFactory.DecodeStream(stream);
                  image = new BitmapDrawable(bitmap);
               }
            }
         }
         catch (FormatException formatException)
         {
            Log.Error(GetType().Name, formatException.Message, formatException);
            throw;
         }
         catch (Exception ex)
         {
            Log.Error(GetType().Name, ex.Message, "Maybe decoding Bitmap stream failed");
#pragma warning disable 618
            image = Resources.GetDrawable(Resource.Drawable.NoQuestion);
#pragma warning restore 618
         }

         return image;
      }

      /// <summary>
      ///    Возвращает строку Uri для изображения по номеру вопроса
      /// </summary>
      /// <param name="questionNumber">Номер вопроса</param>
      /// <returns>Uri-изображения</returns>
      private string GetQuestionImageUri(int questionNumber)
         =>
            _questions.TryGetValue(questionNumber, out Question currentQuestion)
               ? currentQuestion.ImageUri
               : string.Empty;

      /// <summary>
      ///    Возвращает строку с текстом вопроса по его номеру
      /// </summary>
      /// <param name="questionNumber">Номер вопроса</param>
      /// <returns>Текст вопроса</returns>
      private string GetQuestionText(int questionNumber)
         => _questions.TryGetValue(questionNumber, out Question currentQuestion) ? currentQuestion.Text : string.Empty;

      /// <summary>
      ///    Загрузка XML в поле класса
      /// </summary>
      /// <param name="startQuestionNumber">Номер вопроса</param>
      private void LoadQuestionBatch(int startQuestionNumber)
      {
         _questions.Clear();
         using (
            var questionsXmlReader = startQuestionNumber < 16
               ? Resources.GetXml(Resource.Xml.SampleQuestions)
               : Resources.GetXml(Resource.Xml.SampleQuestions2))
         {
            // Ищем записи для score из XML
            while (questionsXmlReader.Read())
            {
               if (questionsXmlReader.NodeType != XmlNodeType.Element)
                  continue;

               // Получаем имя тэга
               var tagName = questionsXmlReader.Name;
               if (tagName == XmlTagQuestion)
               {
                  var questionNumber =
                     int.Parse(questionsXmlReader.MoveToAttribute(XmlTagQuestionAttributeNumber)
                        ? questionsXmlReader.Value
                        : "0");
                  var questionText = questionsXmlReader.MoveToAttribute(XmlTagQuestionAttributeText)
                     ? questionsXmlReader.Value
                     : string.Empty;
                  var questionImageUri = questionsXmlReader.MoveToAttribute(XmlTagQuestionAttributeImageurl)
                     ? questionsXmlReader.Value
                     : string.Empty;
                  _questions.Add(questionNumber, new Question(questionText, questionImageUri));
               }
            }
         }
      }

      /// <summary>
      ///    Вспомогательный метод для записи ответа и загрузки нового вопроса
      /// </summary>
      /// <param name="yesAnswer">true, если пользователь дал положительный ответ, false в противном случае</param>
      private void HandleAnswerAndShowNextQuestion(bool yesAnswer = false)
      {
         var currentScore = _gameSettings.GetInt(GamePreferencesScore, default(int));
         var nextQuestionNumber = _gameSettings.GetInt(GamePreferencesCurrentQuestion, 1) + 1;
         var editor = _gameSettings.Edit();
         editor.PutInt(GamePreferencesCurrentQuestion, nextQuestionNumber);

         // Учитываем ответ в случае Yes
         if (yesAnswer)
            editor.PutInt(GamePreferencesScore, currentScore + 1);

         editor.Commit();

         if (!_questions.ContainsKey(nextQuestionNumber))
            LoadQuestionBatch(nextQuestionNumber);

         if (_questions.ContainsKey(nextQuestionNumber))
         {
            _questionTextSwitcher?.SetText(GetQuestionText(nextQuestionNumber));
            _questionImageSwitcher?.SetImageDrawable(GetQuestionImage(nextQuestionNumber));
         }
         else
         {
            HandleNoQuestions();
         }
      }

      #region Internal types

      /// <summary>
      ///    Класс для управления данными вопроса викторины
      /// </summary>
      private sealed class Question
      {
         public Question(string text, string imageUri)
         {
            Text = text;
            ImageUri = imageUri;
         }

         /// <summary>
         ///    The text for this question
         /// </summary>
         public string Text { get; }

         /// <summary>
         ///    A valid image Url to display with this question
         /// </summary>
         public string ImageUri { get; }
      }

      /// <summary>
      ///    Фабрика переключателя для использования с изображением вопроса. Создает следующий объект для анимации
      /// </summary>
      private sealed class QuizGameTextSwitcherFactory : ViewSwitcher.IViewFactory
      {
         private readonly QuizGameActivity _quizGameActivity;

         public QuizGameTextSwitcherFactory(QuizGameActivity quizGameActivity)
         {
            _quizGameActivity = quizGameActivity;
         }

         public void Dispose()
         {
         }

         // ReSharper disable UnassignedGetOnlyAutoProperty
         public IntPtr Handle { get; }
         // ReSharper restore UnassignedGetOnlyAutoProperty

         public View MakeView()
         {
            var textView = new TextView(_quizGameActivity)
            {
               Gravity = GravityFlags.Center,
               TextSize = _quizGameActivity.Resources.GetDimension(Resource.Dimension.game_question_size)
            };
#pragma warning disable 618
            textView.SetTextColor(_quizGameActivity.Resources.GetColor(Resource.Color.title_color));
            textView.SetShadowLayer(10, 5, 5, _quizGameActivity.Resources.GetColor(Resource.Color.title_glow));
#pragma warning restore 618

            return textView;
         }
      }

      private sealed class QuizGameImageSwitcherFactory : ViewSwitcher.IViewFactory
      {
         private readonly QuizGameActivity _quizGameActivity;

         public QuizGameImageSwitcherFactory(QuizGameActivity quizGameActivity)
         {
            _quizGameActivity = quizGameActivity;
         }

         public void Dispose()
         {
         }

         // ReSharper disable UnassignedGetOnlyAutoProperty
         public IntPtr Handle { get; }
         // ReSharper restore UnassignedGetOnlyAutoProperty

         public View MakeView()
         {
            var imageView = new ImageView(_quizGameActivity)
            {
               LayoutParameters =
                  new FrameLayout.LayoutParams(ViewGroup.LayoutParams.WrapContent, ViewGroup.LayoutParams.WrapContent)
            };
            imageView.SetScaleType(ImageView.ScaleType.FitCenter);

            return imageView;
         }
      }

      #endregion
   }
}

File: - ToCode/Android/TriviaQuiz/QuizGameActivity.java
public class QuizGameActivity extends QuizActivity
{
   /**
    * Вспомогательный метод для записи ответа и загрузки нового вопроса
    *
    * @param bAnswer true, если пользователь дал положительный ответ, false в противном случае
    */
   private void handleAnswerAndShowNextQuestion(boolean bAnswer)
   {
      int curScore = mGameSettings.getInt(GAME_PREFERENCES_SCORE, 0);
      int nextQuestionNumber = mGameSettings.getInt(GAME_PREFERENCES_CURRENT_QUESTION, 1) + 1;

      Editor editor = mGameSettings.edit();
      editor.putInt(GAME_PREFERENCES_CURRENT_QUESTION, nextQuestionNumber);

      // -- Учитываем ответ в случае Yes
      if (bAnswer)
      {
         editor.putInt(GAME_PREFERENCES_SCORE, curScore + 1);
      }
      editor.commit();

      if (!mQuestions.containsKey(nextQuestionNumber))
      {
         try
         {  // -- Загружаем следующую серию вопросов
            loadQuestionBatch(nextQuestionNumber);
         }
         catch (Exception e)
         {
            Log.e(DEBUG_TAG, "Loading updated question batch failed", e);
         }
      }

      if (mQuestions.containsKey(nextQuestionNumber))
      {
         // -- Обновляем текст вопроса
         TextSwitcher questionTextSwitcher =
           (TextSwitcher) findViewById(R.id.TextSwitcher_QuestionText);
         questionTextSwitcher.setText(getQuestionText(nextQuestionNumber));

         // -- Обновляем изображение вопроса
         ImageSwitcher questionImageSwitcher =
           (ImageSwitcher) findViewById(R.id.ImageSwitcher_QuestionImage);
         Drawable image = getQuestionImageDrawable(nextQuestionNumber);
         questionImageSwitcher.setImageDrawable(image);
      }
      else
      {
         handleNoQuestions();
      }

   }

   /**
    * Вспомогательный метод, отвечающий за ситуацию отсутствия вопросов
    */
   private void handleNoQuestions()
   {
      TextSwitcher questionTextSwitcher =
        (TextSwitcher) findViewById(R.id.TextSwitcher_QuestionText);
      questionTextSwitcher.setText(
        getResources().getText(R.string.no_questions));
      ImageSwitcher questionImageSwitcher =
        (ImageSwitcher) findViewById(R.id.ImageSwitcher_QuestionImage);
      questionImageSwitcher.setImageResource(R.drawable.noquestion);

      // -- Отключаем кнопку Yes
      Button yesButton = (Button) findViewById(R.id.Button_Yes);
      yesButton.setEnabled(false);

      // -- Отключаем кнопку No
      Button noButton = (Button) findViewById(R.id.Button_No);
      noButton.setEnabled(false);
   }

   /**
    * Возвращает строку с текстом вопроса по его номеру
    *
    * @param questionNumber Номер вопроса
    * @return Текст вопроса
    */
   private String getQuestionText(Integer questionNumber)
   {
      String text = null;
      Question curQuestion = mQuestions.get(questionNumber);
      if (curQuestion != null)
      {
         text = curQuestion.getmText();
      }
      return text;
   }

   /**
    * Возвращает строку URL для изображения по номеру вопроса
    *
    * @param questionNumber Номер вопроса
    * @return URL изображения
    */
   private String getQuestionImageUrl(Integer questionNumber)
   {
      String url = null;
      Question curQuestion = mQuestions.get(questionNumber);
      if (curQuestion != null)
      {
         url = curQuestion.getmImageUrl();
      }
      return url;
   }

   /**
    * Извлекает объект Drawable для вопроса
    *
    * @param questionNumber Номер вопроса
    * @return Объект Drawable
    */
   @SuppressWarnings("deprecation")
   private Drawable getQuestionImageDrawable(int questionNumber)
   {
      Drawable image;
      URL imageUrl;

      try
      {
         // -- Создание объекта Drawable через декодирование потока из URL
         imageUrl = new URL(getQuestionImageUrl(questionNumber));
         Bitmap bitmap = BitmapFactory.decodeStream(imageUrl.openStream());
         image = new BitmapDrawable(bitmap);
      }
      catch (Exception e)
      {
         Log.e(DEBUG_TAG, "Decoding Bitmap stream failed.");
         image = getResources().getDrawable(R.drawable.noquestion);
      }
      return image;
   }

   /**
    * Загрузка XML в поле класса
    *
    * @param startQuestionNumber TODO: Пока не используется
    * @throws XmlPullParserException
    * @throws IOException
    */
   private void loadQuestionBatch(int startQuestionNumber) throws XmlPullParserException, IOException
   {
      // Удаляем старую серию вопросов
      mQuestions.clear();

      // TODO: Устанавливаем связь с сервером и извлекаем новую серию вопросов
      XmlResourceParser questionBatch;

      questionBatch = startQuestionNumber < 16
        ? getResources().getXml(R.xml.samplequestions)
        : getResources().getXml(R.xml.samplequestions2);

      int eventType = -1;

      // Ищем записи для score из XML
      while (eventType != XmlResourceParser.END_DOCUMENT)
      {
         if (eventType == XmlResourceParser.START_TAG)
         {
            // Получаем имя тэга
            String strName = questionBatch.getName();

            if (strName.equals(XML_TAG_QUESTION))
            {
               String questionNumber = questionBatch.
                 getAttributeValue(null, XML_TAG_QUESTION_ATTRIBUTE_NUMBER);
               Integer questionNum = Integer.valueOf(questionNumber);
               String questionText =
                 questionBatch.getAttributeValue(null, XML_TAG_QUESTION_ATTRIBUTE_TEXT);
               String questionImageUrl =
                 questionBatch.getAttributeValue(null, XML_TAG_QUESTION_ATTRIBUTE_IMAGEURL);

               // Сохранение данных с карту
               mQuestions.put(
                 questionNum, new Question(questionNum, questionText, questionImageUrl));
            }
         }
         eventType = questionBatch.next();
      }
   }   
}
